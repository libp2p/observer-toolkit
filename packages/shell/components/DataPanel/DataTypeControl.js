import React, { useState } from 'react'
import T from 'prop-types'
import styled from 'styled-components'

import { Icon } from '@libp2p-observer/sdk'

const Container = styled.div`
  display: flex;
  background: ${({ theme }) => theme.color('contrast', 1)};
  cursor: pointer;
  ${({ theme }) => theme.text('label', 'small')}
`

const IconButton = styled.button`
  background: ${({ theme }) => theme.color('background')};
  margin: ${({ theme }) => theme.spacing([0, 1, 0, 0.5])};
  color: ${({ theme }) => theme.color('highlight')};
  border-radius: 50%;
  border: none;
  cursor: pointer;
  position: relative;
  padding: ${({ theme }) => theme.spacing(0.5)};
  z-index: 2;
`

const ButtonText = styled.span`
  flex-grow: 1;
  background: ${({ theme }) => theme.color('contrast', 2)};
  border-radius: ${({ theme }) => theme.spacing()};
  color: ${({ theme }) => theme.color('text', 3)};
  padding: ${({ theme }) => theme.spacing([0.5, 1, 0.5, 2.5])};
  margin: ${({ theme }) => theme.spacing([0.5, 0, 0.5, -2.5])};
  text-align: left;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  ${({ theme, isHighlighted }) =>
    isHighlighted &&
    `
    background: ${theme.color('background')};
    color: ${theme.color('highlight', 1)};
  `}
`

function DataTypeControl({ metadata, openDataTray }) {
  const [isHighlighted, setHighlighted] = useState(false)

  if (!metadata) return ''

  const { type, name } = metadata

  const iconNames = {
    sample: 'cloud',
    upload: 'doc',
  }
  if (!iconNames[type]) {
    throw new Error(
      `Unknown type "${type}", expected one of "${Object.keys(iconNames).join(
        '", "'
      )}"`
    )
  }

  const iconType = isHighlighted ? 'back' : iconNames[type]

  // TODO: When implementing live ws mode:
  //  - Make icon pulse gently while live
  //  - Show "pause" icon when paused
  return (
    <Container
      onMouseEnter={() => setHighlighted(true)}
      onMouseLeave={() => setHighlighted(false)}
      data-highlighted={isHighlighted}
      onClick={openDataTray}
    >
      <IconButton isHighlighted={isHighlighted}>
        <Icon type={iconType} />
      </IconButton>
      <ButtonText isHighlighted={isHighlighted}>
        {isHighlighted ? 'Change data source' : name}
      </ButtonText>
    </Container>
  )
}

DataTypeControl.propTypes = {
  metadata: T.object,
  openDataTray: T.func.isRequired,
}

export default DataTypeControl
