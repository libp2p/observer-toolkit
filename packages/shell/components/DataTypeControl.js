import React, { useState } from 'react'
import T from 'prop-types'
import styled from 'styled-components'

import { Icon } from '@libp2p-observer/sdk'

const Container = styled.div`
  display: flex;
  background: ${({ theme }) => theme.color('contrast', 1)};
  cursor: pointer;
  ${({ theme }) => theme.text('label', 'small')}
`

const IconButton = styled.button`
  background: ${({ theme }) => theme.color('background')};
  margin-right: ${({ theme }) => theme.spacing(1)};
  color: ${({ theme }) => theme.color('highlight')};
  border-radius: 50%;
  border: none;
  cursor: pointer;
  position: relative;
  z-index: 2;
`

const ButtonText = styled.span`
  flex-grow: 1;
  background: ${({ theme }) => theme.color('contrast', 2)};
  border-radius: ${({ theme }) => theme.spacing()};
  color: ${({ theme }) => theme.color('text', 2)};
  padding-top: ${({ theme }) => theme.spacing(0.5)};
  padding-bottom: ${({ theme }) => theme.spacing(0.5)};
  padding-left: ${({ theme }) => theme.spacing(2.5)};
  padding-right: ${({ theme }) => theme.spacing(1)};
  margin-left: ${({ theme }) => theme.spacing(-2.5)};
  margin-top: ${({ theme }) => theme.spacing(0.5)};
  margin-bottom: ${({ theme }) => theme.spacing(0.5)};
  text-align: left;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  ${({ theme, isHighlighted }) =>
    isHighlighted &&
    `
    background: ${theme.color('background')};
    color: ${theme.color('highlight', 1)};
  `}
`

function DataTypeControl({ metadata }) {
  const [isHighlighted, setHighlighted] = useState(false)

  if (!metadata) return ''

  const { type, name } = metadata

  const iconNames = {
    sample: 'cloud',
    upload: 'doc',
  }

  const iconType = isHighlighted ? 'back' : iconNames[type]

  // TODO: When implementing live ws mode:
  //  - Make icon pulse gently while live
  //  - Show "pause" icon when paused
  return (
    <Container
      onMouseEnter={() => setHighlighted(true)}
      onMouseLeave={() => setHighlighted(false)}
    >
      <IconButton isHighlighted={isHighlighted}>
        <Icon type={iconType} />
      </IconButton>
      <ButtonText isHighlighted={isHighlighted}>
        {isHighlighted ? 'Change data source' : name}
      </ButtonText>
    </Container>
  )
}

DataTypeControl.propTypes = {
  metadata: T.object,
}

export default DataTypeControl
