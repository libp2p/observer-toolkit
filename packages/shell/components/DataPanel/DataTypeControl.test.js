import React from 'react'
import T from 'prop-types'

import { fireEvent } from '@testing-library/react'

import {
  catchErrorSilently,
  renderWithTheme,
  getMockRuntime,
} from '@nearform/observer-testing'
import { DataProvider } from '@nearform/observer-sdk'
import DataTypeControl from './DataTypeControl'

const initialData = {
  runtime: getMockRuntime(),
}

function DataTypeWithSource({ source, openDataTray = () => {} }) {
  return (
    <DataProvider initialSource={source} initialData={initialData}>
      <DataTypeControl openDataTray={openDataTray} />
    </DataProvider>
  )
}
DataTypeWithSource.propTypes = {
  source: T.object.isRequired,
  openDataTray: T.any,
}

describe('DataTypeControl', () => {
  it('renders as expected for sample data', () => {
    const source = {
      type: 'sample',
      name: 'test sample name',
    }
    const { asFragment } = renderWithTheme(
      <DataTypeWithSource source={source} />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders as expected for uploaded file', () => {
    const source = {
      type: 'upload',
      name: 'test filename',
    }
    const { asFragment } = renderWithTheme(
      <DataTypeWithSource source={source} />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('throws if trying to render an invalid type', () => {
    const source = {
      type: 'invalid-type',
      name: 'test invalid type',
    }
    expect(
      catchErrorSilently(() =>
        renderWithTheme(<DataTypeWithSource source={source} />)
      )
    ).toBeInstanceOf(Error)
  })

  it('is highlighted and text changes on mouseover', async () => {
    const name = 'test mouseover target'
    const highlightedText = 'Change data source'

    const source = {
      type: 'sample',
      name,
    }
    const {
      getByText,
      queryAllByText,
      queryAllByHighlighted,
    } = renderWithTheme(<DataTypeWithSource source={source} />)
    expect(queryAllByHighlighted()).toHaveLength(0)
    expect(queryAllByText(highlightedText)).toHaveLength(0)

    await fireEvent.mouseEnter(getByText(name))
    expect(queryAllByHighlighted()).toHaveLength(1)
    expect(queryAllByText(highlightedText)).toHaveLength(1)
    expect(queryAllByText(name)).toHaveLength(0)

    await fireEvent.mouseLeave(getByText(highlightedText))
    expect(queryAllByHighlighted()).toHaveLength(0)
    expect(queryAllByText(highlightedText)).toHaveLength(0)
    expect(queryAllByText(name)).toHaveLength(1)
  })
})
