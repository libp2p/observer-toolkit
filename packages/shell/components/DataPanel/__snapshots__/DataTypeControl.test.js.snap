// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTypeControl renders as expected for sample data 1`] = `
<DocumentFragment>
  .c3 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  display: inline-block;
  vertical-align: middle;
  width: 2em;
  height: 2em;
}

.c0 {
  width: inherit;
  height: inherit;
  display: inherit;
  -webkit-flex-direction: inherit;
  -ms-flex-direction: inherit;
  flex-direction: inherit;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background: rgb(78,23,91);
  cursor: pointer;
  font-family: plex-sans,sans-serif;
  line-height: 1.6em;
  font-size: 8pt;
  line-height: 1.2em;
}

.c2 {
  background: rgb(255,255,255);
  margin: 0px 8px 0px 4px;
  color: rgb(209,14,102);
  border-radius: 50%;
  border: none;
  cursor: pointer;
  position: relative;
  padding: 4px;
  z-index: 2;
}

.c4 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  background: rgb(121,27,137);
  border-radius: 8px;
  color: rgb(255,255,255);
  padding: 4px 8px 4px 20px;
  margin: 4px 0px 4px -20px;
  text-align: left;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

<div
    class="c0"
  >
    <div
      class="c1"
      data-highlighted="false"
    >
      <button
        class="c2"
      >
        <span
          class="c3"
          role="img"
        >
          <svg
            viewBox="0 0 96.62 96.62"
          >
            <path
              d="M72.84 44.9a9.67 9.67 0 00-4.57 1.1 10.88 10.88 0 00.06-1.14 10.7 10.7 0 00-10.57-10.8A10.45 10.45 0 0049.64 38a13.31 13.31 0 00-12.25-8.31A13.46 13.46 0 0024.1 42.58a11.35 11.35 0 00-9.83 11.35c0 11.65 10 13.05 10.52 13.05H72.7c3.47 0 9.35-5.53 9.35-11.18s-3.69-10.9-9.21-10.9z"
              data-name="Layer 40"
              fill="none"
              stroke="currentColor"
              stroke-miterlimit="10"
              stroke-width="4"
            />
          </svg>
        </span>
      </button>
      <span
        class="c4"
      >
        test sample name
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`DataTypeControl renders as expected for uploaded file 1`] = `
<DocumentFragment>
  .c3 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  display: inline-block;
  vertical-align: middle;
  width: 2em;
  height: 2em;
}

.c0 {
  width: inherit;
  height: inherit;
  display: inherit;
  -webkit-flex-direction: inherit;
  -ms-flex-direction: inherit;
  flex-direction: inherit;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background: rgb(78,23,91);
  cursor: pointer;
  font-family: plex-sans,sans-serif;
  line-height: 1.6em;
  font-size: 8pt;
  line-height: 1.2em;
}

.c2 {
  background: rgb(255,255,255);
  margin: 0px 8px 0px 4px;
  color: rgb(209,14,102);
  border-radius: 50%;
  border: none;
  cursor: pointer;
  position: relative;
  padding: 4px;
  z-index: 2;
}

.c4 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  background: rgb(121,27,137);
  border-radius: 8px;
  color: rgb(255,255,255);
  padding: 4px 8px 4px 20px;
  margin: 4px 0px 4px -20px;
  text-align: left;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

<div
    class="c0"
  >
    <div
      class="c1"
      data-highlighted="false"
    >
      <button
        class="c2"
      >
        <span
          class="c3"
          role="img"
        >
          <svg
            viewBox="0 0 96 96"
          >
            <defs>
              <style>
                .doc_svg__cls-2{fill:none;stroke:currentColor;stroke-miterlimit:10;stroke-width:4px}
              </style>
            </defs>
            <g
              data-name="Layer 25"
              id="doc_svg__Layer_25"
            >
              <path
                d="M66.53 18.76a8 8 0 018 8v42.48a8 8 0 01-8 8H29.47a8 8 0 01-8-8V26.76a8 8 0 018-8h37.06m0-4H29.47a12 12 0 00-12 12v42.48a12 12 0 0012 12h37.06a12 12 0 0012-12V26.76a12 12 0 00-12-12z"
                fill="currentColor"
              />
              <path
                class="doc_svg__cls-2"
                d="M30.94 32.81h34.12M30.94 45h34.12M30.94 57.2h34.12"
              />
            </g>
          </svg>
        </span>
      </button>
      <span
        class="c4"
      >
        test filename
      </span>
    </div>
  </div>
</DocumentFragment>
`;
